internal/wazeroir/compiler.go

```diff
@@ -2019,889 +2019,890 @@ operatorSwitch:
 			c.pc++
 			laneIndex := c.body[c.pc]
 			c.emit(
-				OperationV128ExtractLane{LaneIndex: laneIndex, Shape: ShapeF32x4},
+				NewOperationV128ExtractLane(laneIndex, false, ShapeF32x4),
 			)
 		case wasm.OpcodeVecF64x2ExtractLane:
 			c.pc++
 			laneIndex := c.body[c.pc]
 			c.emit(
-				OperationV128ExtractLane{LaneIndex: laneIndex, Shape: ShapeF64x2},
+				NewOperationV128ExtractLane(laneIndex, false, ShapeF64x2),
 			)
 		case wasm.OpcodeVecI8x16ReplaceLane:
 			c.pc++
 			laneIndex := c.body[c.pc]
 			c.emit(
-				OperationV128ReplaceLane{LaneIndex: laneIndex, Shape: ShapeI8x16},
+				NewOperationV128ReplaceLane(laneIndex, ShapeI8x16),
 			)
 		case wasm.OpcodeVecI16x8ReplaceLane:
 			c.pc++
 			laneIndex := c.body[c.pc]
 			c.emit(
-				OperationV128ReplaceLane{LaneIndex: laneIndex, Shape: ShapeI16x8},
+				NewOperationV128ReplaceLane(laneIndex, ShapeI16x8),
 			)
 		case wasm.OpcodeVecI32x4ReplaceLane:
 			c.pc++
 			laneIndex := c.body[c.pc]
 			c.emit(
-				OperationV128ReplaceLane{LaneIndex: laneIndex, Shape: ShapeI32x4},
+				NewOperationV128ReplaceLane(laneIndex, ShapeI32x4),
 			)
 		case wasm.OpcodeVecI64x2ReplaceLane:
 			c.pc++
 			laneIndex := c.body[c.pc]
 			c.emit(
-				OperationV128ReplaceLane{LaneIndex: laneIndex, Shape: ShapeI64x2},
+				NewOperationV128ReplaceLane(laneIndex, ShapeI64x2),
 			)
 		case wasm.OpcodeVecF32x4ReplaceLane:
 			c.pc++
 			laneIndex := c.body[c.pc]
 			c.emit(
-				OperationV128ReplaceLane{LaneIndex: laneIndex, Shape: ShapeF32x4},
+				NewOperationV128ReplaceLane(laneIndex, ShapeF32x4),
 			)
 		case wasm.OpcodeVecF64x2ReplaceLane:
 			c.pc++
 			laneIndex := c.body[c.pc]
 			c.emit(
-				OperationV128ReplaceLane{LaneIndex: laneIndex, Shape: ShapeF64x2},
+				NewOperationV128ReplaceLane(laneIndex, ShapeF64x2),
 			)
 		case wasm.OpcodeVecI8x16Splat:
 			c.emit(
-				OperationV128Splat{Shape: ShapeI8x16},
+				NewOperationV128Splat(ShapeI8x16),
 			)
 		case wasm.OpcodeVecI16x8Splat:
 			c.emit(
-				OperationV128Splat{Shape: ShapeI16x8},
+				NewOperationV128Splat(ShapeI16x8),
 			)
 		case wasm.OpcodeVecI32x4Splat:
 			c.emit(
-				OperationV128Splat{Shape: ShapeI32x4},
+				NewOperationV128Splat(ShapeI32x4),
 			)
 		case wasm.OpcodeVecI64x2Splat:
 			c.emit(
-				OperationV128Splat{Shape: ShapeI64x2},
+				NewOperationV128Splat(ShapeI64x2),
 			)
 		case wasm.OpcodeVecF32x4Splat:
 			c.emit(
-				OperationV128Splat{Shape: ShapeF32x4},
+				NewOperationV128Splat(ShapeF32x4),
 			)
 		case wasm.OpcodeVecF64x2Splat:
 			c.emit(
-				OperationV128Splat{Shape: ShapeF64x2},
+				NewOperationV128Splat(ShapeF64x2),
 			)
 		case wasm.OpcodeVecI8x16Swizzle:
 			c.emit(
-				OperationV128Swizzle{},
+				NewOperationV128Swizzle(),
 			)
 		case wasm.OpcodeVecV128i8x16Shuffle:
 			c.pc++
-			op := OperationV128Shuffle{}
-			copy(op.Lanes[:], c.body[c.pc:c.pc+16])
+			lanes := [16]byte{}
+			copy(lanes[:], c.body[c.pc:c.pc+16])
+			op := NewOperationV128Shuffle(lanes)
```

Comment:
is this initialization done right?

Reply:
left a comment in the impl of NewOperationV128Shuffle.

Reply:
if we don't do a defensive copy here we're going to share the subrange in the slice (?). Is this ok? (I assume so because we I don't think we make changes in-place, but...)

Reply:
no in any way you end up copying. please try my suggestion in the comment.

Reply:
basically, you are copying twice in your current code.